File: myflux-lib/src/main/java/com/github/bluzwong/myflux/lib/BaseFluxActivity.java
package com.github.bluzwong.myflux.lib;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;

import java.util.Map;

/**
 * Created by wangzhijie@wind-mobi.com on 2015/9/25.
 */
public abstract class BaseFluxActivity extends AppCompatActivity {
    protected abstract int provideContentId();

    protected abstract FluxStore provideStore();

    protected abstract void init();

    protected abstract void onRestoreView();

    protected abstract void onNewView();

    protected abstract void onResponse(String type, Map<String, Object> dataMap);

    private int hashCode = hashCode();
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(provideContentId());
        init();
        FluxStore store = provideStore();
        if (store == null || !store.register(this, savedInstanceState, new Runnable() {
            @Override
            public void run() {
                onRestoreView();
            }
        })) {
            onNewView();
        }
    }

    public void onReceiveResponse(FluxResponse response) {
        if (response.getOwner() != hashCode) {
            return;
        }
        onResponse(response.getType(), response.getDataMap());
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        if (provideStore() != null) {
            provideStore().onSaveInstanceState(outState);
        }
    }

    @Override
    protected void onDestroy() {
        if (provideStore() != null) {
            provideStore().unregister(this);
        }
        super.onDestroy();
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (provideStore() != null) {
            provideStore().onResume();
        }
    }
}

File: myflux-lib/src/main/kotlin/com/github/bluzwong/myflux/lib/kotlin/BaseFluxActivity.kt
package com.github.bluzwong.myflux.lib.kotlin

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
/**
 * Created by wangzhijie@wind-mobi.com on 2015/9/15.
 */
public abstract class BaseFluxActivity : android.support.v7.app.AppCompatActivity(), FluxResponse {

    /**
     * provide layout resource id
     */
    protected abstract fun provideContentId(): Int

    /**
     * provide store to hold and control view state and data
     */
    protected abstract fun provideStore(): Store?

    /**
     * init views
     */
    protected abstract fun init():Unit

    /**
     * called when activity is being restored
     * check the info in store and refresh the view
     */
    protected abstract fun onRestoreView():Unit

    /**
     * called when is the first time showing
     */
    protected abstract fun onNewView():Unit

    override fun onCreate(savedInstanceState: android.os.Bundle?) {
        super<AppCompatActivity>.onCreate(savedInstanceState)
        setContentView(provideContentId())
        init()
        if (provideStore()?.register(this, savedInstanceState, { onRestoreView() })?:false) {

        } else {
            onNewView()
        }
    }

    override public fun onSaveInstanceState(outState: android.os.Bundle?) {
        super<AppCompatActivity>.onSaveInstanceState(outState)
        provideStore()?.onSaveInstanceState(outState ?: return)
    }

    override protected fun onDestroy() {
        provideStore()?.unregister(this)
        super<AppCompatActivity>.onDestroy()
    }

    override protected fun onResume() {
        super<AppCompatActivity>.onResume()
        provideStore()?.onResume()
    }
}